File: src/Todo/Sync/CRDT.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}

module Todo.Sync.CRDT
  ( -- * Operation Application
    applyOperation
  , applyOperations
    -- * Operation Merging
  , mergeOperations
  , deduplicateOperations
    -- * State Materialization
  , materialize
  , materializeToTodos
    -- * Operation Generation
  , createAddOp
  , createCompleteOp
  , createUncompleteOp
  , createDeleteOp
  , createSetPriorityOp
  , createModifyDescriptionOp
    -- * Diff Detection
  , diffToOperations
  , itemFingerprint
    -- * Utilities
  , syncItemToTodo
  , filterPendingOps
  ) where

import           RIO
import qualified RIO.Text as T

import qualified Crypto.Hash.SHA256    as SHA256
import qualified Data.ByteString.Base16 as B16
import qualified Data.Map.Strict       as Map
import           Data.List             (sortBy, nubBy)
import           Data.Time.Clock       (UTCTime)
import qualified Data.Text.Encoding    as TE

import           Todo.Types            (Priority, TodoItem(..), Todo(..), Metadata)
import           Todo.Sync.Types

-- | Apply a single operation to the sync state
applyOperation :: SyncState -> Operation -> SyncState
applyOperation state op = state { ssItems = updatedItems, ssOperations = op : ssOperations state }
  where
    updatedItems = case op of
      OpAdd{..} ->
        let newItem = SyncItem
              { siItemId      = opItemId
              , siItem        = opItem
              , siCompleted   = False
              , siDeleted     = False
              , siPriority    = LWWRegister (tPriority opItem) opTimestamp opDeviceId
              , siDescription = LWWRegister (tDescription opItem) opTimestamp opDeviceId
              , siCreatedAt   = opTimestamp
              }
        in Map.insertWith mergeItems opItemId newItem (ssItems state)

      OpComplete{..} ->
        Map.adjust (\item -> item { siCompleted = True }) opItemId (ssItems state)

      OpUncomplete{..} ->
        Map.adjust (\item -> item { siCompleted = False }) opItemId (ssItems state)

      OpDelete{..} ->
        Map.adjust (\item -> item { siDeleted = True }) opItemId (ssItems state)

      OpSetPriority{..} ->
        Map.adjust (updatePriority opPriority opTimestamp opDeviceId) opItemId (ssItems state)

      OpModifyDescription{..} ->
        Map.adjust (updateDescription opDescription opTimestamp opDeviceId) opItemId (ssItems state)

    -- When adding an item that already exists, merge using LWW for conflicting fields
    mergeItems new old = old
      { siPriority    = mergeLWW (siPriority new) (siPriority old)
      , siDescription = mergeLWW (siDescription new) (siDescription old)
      -- Complete-wins: if either is completed, stay completed
      , siCompleted   = siCompleted new || siCompleted old
      -- Delete-wins: if either is deleted, stay deleted
      , siDeleted     = siDeleted new || siDeleted old
      }

    updatePriority pri timestamp deviceId item =
      let newReg = LWWRegister pri timestamp deviceId
      in item { siPriority = mergeLWW newReg (siPriority item) }

    updateDescription desc timestamp deviceId item =
      let newReg = LWWRegister desc timestamp deviceId
      in item { siDescription = mergeLWW newReg (siDescription item) }

-- | Apply multiple operations to sync state
applyOperations :: SyncState -> [Operation] -> SyncState
applyOperations = foldl' applyOperation

-- | Merge two operation lists, removing duplicates by operation ID
-- Operations are commutative and idempotent
mergeOperations :: [Operation] -> [Operation] -> [Operation]
mergeOperations ops1 ops2 = deduplicateOperations (ops1 ++ ops2)

-- | Remove duplicate operations based on operation ID
deduplicateOperations :: [Operation] -> [Operation]
deduplicateOperations = nubBy (\a b -> opId a == opId b)

-- | Materialize current state from an operation log
-- Starts from empty state and applies all operations in order
materialize :: DeviceId -> [Operation] -> Map.Map ItemId SyncItem
materialize deviceId ops =
  let sortedOps = sortBy compareByTime ops
      emptyState = emptySyncState deviceId
      finalState = applyOperations emptyState sortedOps
  in ssItems finalState
  where
    compareByTime a b = compare (operationTimestamp a) (operationTimestamp b)

-- | Convert materialized state to list of Todo items (excluding deleted)
materializeToTodos :: Map.Map ItemId SyncItem -> [Todo TodoItem]
materializeToTodos items =
  [ syncItemToTodo item
  | item <- Map.elems items
  , not (siDeleted item)
  ]

-- | Convert a SyncItem to a Todo TodoItem
syncItemToTodo :: SyncItem -> Todo TodoItem
syncItemToTodo SyncItem{..} =
  let item = siItem
        { tPriority    = lwwValue siPriority
        , tDescription = lwwValue siDescription
        }
  in if siCompleted
     then Completed item
     else Incomplete item

-- | Create an Add operation
createAddOp :: DeviceId -> ItemId -> TodoItem -> UTCTime -> IO Operation
createAddOp deviceId itemId item timestamp = do
  opId' <- generateOperationId
  return $ OpAdd opId' itemId item timestamp deviceId

-- | Create a Complete operation
createCompleteOp :: DeviceId -> ItemId -> UTCTime -> IO Operation
createCompleteOp deviceId itemId timestamp = do
  opId' <- generateOperationId
  return $ OpComplete opId' itemId timestamp deviceId

-- | Create an Uncomplete operation
createUncompleteOp :: DeviceId -> ItemId -> UTCTime -> IO Operation
createUncompleteOp deviceId itemId timestamp = do
  opId' <- generateOperationId
  return $ OpUncomplete opId' itemId timestamp deviceId

-- | Create a Delete operation
createDeleteOp :: DeviceId -> ItemId -> UTCTime -> IO Operation
createDeleteOp deviceId itemId timestamp = do
  opId' <- generateOperationId
  return $ OpDelete opId' itemId timestamp deviceId

-- | Create a SetPriority operation
createSetPriorityOp :: DeviceId -> ItemId -> Maybe Priority -> UTCTime -> IO Operation
createSetPriorityOp deviceId itemId priority timestamp = do
  opId' <- generateOperationId
  return $ OpSetPriority opId' itemId priority timestamp deviceId

-- | Create a ModifyDescription operation
createModifyDescriptionOp :: DeviceId -> ItemId -> T.Text -> UTCTime -> IO Operation
createModifyDescriptionOp deviceId itemId description timestamp = do
  opId' <- generateOperationId
  return $ OpModifyDescription opId' itemId description timestamp deviceId

-- | Compute a fingerprint for a todo item based on its content
-- Used for change detection when external edits occur
itemFingerprint :: Todo TodoItem -> T.Text
itemFingerprint todo =
  let item = case todo of
        Completed i  -> i
        Incomplete i -> i
      content = T.concat
        [ tDescription item
        , T.pack $ show $ tMetadata item
        ]
      hash = B16.encode $ SHA256.hash $ TE.encodeUtf8 content
  in TE.decodeUtf8 hash

-- | Diff old vs new todo lists to generate operations
-- This is used when detecting external file changes
diffToOperations
  :: DeviceId
  -> UTCTime
  -> Map.Map ItemId SyncItem   -- ^ Current sync state (with ItemIds)
  -> [Todo TodoItem]            -- ^ Old todos from file
  -> [Todo TodoItem]            -- ^ New todos from file
  -> IO [Operation]
diffToOperations deviceId timestamp currentState oldTodos newTodos = do
  let oldFingerprints = Map.fromList [(itemFingerprint t, t) | t <- oldTodos]
      newFingerprints = Map.fromList [(itemFingerprint t, t) | t <- newTodos]

      -- Items that exist in new but not old (added)
      addedFingerprints = Map.keys $ Map.difference newFingerprints oldFingerprints

      -- Items that exist in old but not new (removed or modified)
      removedFingerprints = Map.keys $ Map.difference oldFingerprints newFingerprints

  -- For added items, create OpAdd operations
  addOps <- forM addedFingerprints $ \fp -> do
    case Map.lookup fp newFingerprints of
      Just todo -> do
        let item = case todo of
              Completed i  -> i
              Incomplete i -> i
        itemId <- generateItemId
        createAddOp deviceId itemId item timestamp
      Nothing -> error "Impossible: fingerprint not found"

  -- For removed items, we need to find the corresponding ItemId
  -- and create appropriate operations (delete or complete)
  deleteOps <- catMaybes <$> forM removedFingerprints (\fp -> do
    case Map.lookup fp oldFingerprints of
      Just oldTodo -> do
        -- Find the ItemId for this item in our sync state
        let matchingItems = [ (itemId, si)
                            | (itemId, si) <- Map.toList currentState
                            , itemFingerprint (syncItemToTodo si) == fp
                            ]
        case matchingItems of
          ((itemId, _):_) -> Just <$> createDeleteOp deviceId itemId timestamp
          [] -> return Nothing  -- Item not in sync state, ignore
      Nothing -> return Nothing)

  -- Check for completion status changes
  completionOps <- catMaybes <$> forM (Map.toList currentState) (\(itemId, si) -> do
    let fp = itemFingerprint (syncItemToTodo si)
        oldCompleted = case Map.lookup fp oldFingerprints of
          Just (Completed _)  -> True
          Just (Incomplete _) -> False
          Nothing             -> siCompleted si
        newCompleted = case Map.lookup fp newFingerprints of
          Just (Completed _)  -> True
          Just (Incomplete _) -> False
          Nothing             -> siCompleted si

    if oldCompleted /= newCompleted
      then if newCompleted
           then Just <$> createCompleteOp deviceId itemId timestamp
           else Just <$> createUncompleteOp deviceId itemId timestamp
      else return Nothing)

  return $ addOps ++ deleteOps ++ completionOps

-- | Filter operations that haven't been synced yet
filterPendingOps :: Maybe UTCTime -> [Operation] -> [Operation]
filterPendingOps Nothing ops = ops
filterPendingOps (Just lastSync) ops =
  filter (\op -> operationTimestamp op > lastSync) ops

File: android/core/src/main/kotlin/com/denibertovic/todo/core/crdt/Crdt.kt
package com.denibertovic.todo.core.crdt

import com.denibertovic.todo.core.types.DeviceId
import com.denibertovic.todo.core.types.ItemId
import com.denibertovic.todo.core.types.Todo

/**
 * Apply a single operation to a [SyncState] and return the new state.
 *
 * Direct translation of `applyOperation` at
 * `src/Todo/Sync/CRDT.hs:43-94`. Preserves all of the CRDT's merge
 * rules:
 *
 * - **Add** merges conflicting fields via LWW on priority and
 *   description, and respects delete-wins and complete-wins on the
 *   boolean flags.
 * - **Complete** flips `completed` to true (never backtracks on its
 *   own — only a later [Operation.Uncomplete] can).
 * - **Uncomplete** flips `completed` to false.
 * - **Delete** flips `deleted` to true and is never reversed.
 * - **SetPriority** and **ModifyDescription** wrap the new value in
 *   an LWW register and merge against the current one.
 *
 * Operations targeting item IDs that have never been `Add`ed are
 * ignored (matching Haskell's `Map.adjust` semantics, which leaves
 * the map unchanged if the key is missing).
 */
fun SyncState.apply(op: Operation): SyncState {
    val newItems: Map<ItemId, SyncItem> = when (op) {
        is Operation.Add -> {
            val newItem = SyncItem(
                itemId = op.itemId,
                item = op.item,
                completed = false,
                deleted = false,
                priority = LWWRegister(op.item.priority, op.timestamp, op.deviceId),
                description = LWWRegister(op.item.description, op.timestamp, op.deviceId),
                createdAt = op.timestamp,
            )
            val existing = items[op.itemId]
            if (existing == null) {
                items + (op.itemId to newItem)
            } else {
                // Conflicting Add — merge LWW registers, respecting
                // delete-wins and complete-wins on booleans.
                val merged = existing.copy(
                    priority = newItem.priority.merge(existing.priority),
                    description = newItem.description.merge(existing.description),
                    completed = newItem.completed || existing.completed,
                    deleted = newItem.deleted || existing.deleted,
                )
                items + (op.itemId to merged)
            }
        }

        is Operation.Complete -> items[op.itemId]?.let { item ->
            items + (op.itemId to item.copy(completed = true))
        } ?: items

        is Operation.Uncomplete -> items[op.itemId]?.let { item ->
            items + (op.itemId to item.copy(completed = false))
        } ?: items

        is Operation.Delete -> items[op.itemId]?.let { item ->
            items + (op.itemId to item.copy(deleted = true))
        } ?: items

        is Operation.SetPriority -> items[op.itemId]?.let { item ->
            val newReg = LWWRegister(op.priority, op.timestamp, op.deviceId)
            items + (op.itemId to item.copy(priority = newReg.merge(item.priority)))
        } ?: items

        is Operation.ModifyDescription -> items[op.itemId]?.let { item ->
            val newReg = LWWRegister(op.description, op.timestamp, op.deviceId)
            items + (op.itemId to item.copy(description = newReg.merge(item.description)))
        } ?: items
    }

    return copy(items = newItems, operations = listOf(op) + operations)
}

/** Apply a batch of operations in order. */
fun SyncState.applyAll(ops: List<Operation>): SyncState =
    ops.fold(this) { st, op -> st.apply(op) }

/**
 * Deduplicate operations by [Operation.opId]. Operations are
 * commutative and idempotent, so dropping duplicates is always safe.
 * Mirrors `deduplicateOperations` at `src/Todo/Sync/CRDT.hs:102-103`.
 */
fun List<Operation>.dedupe(): List<Operation> {
    val seen = HashSet<com.denibertovic.todo.core.types.OperationId>(size)
    val result = ArrayList<Operation>(size)
    for (op in this) {
        if (seen.add(op.opId)) result += op
    }
    return result
}

/** Merge two operation lists, dropping duplicates. */
fun mergeOperations(a: List<Operation>, b: List<Operation>): List<Operation> =
    (a + b).dedupe()

/**
 * Fold an operation log into a materialized item map, starting from
 * empty state. Operations are sorted by timestamp first, so the
 * result is independent of delivery order — that's what makes the
 * CRDT commutative.
 *
 * Mirrors `materialize` at `src/Todo/Sync/CRDT.hs:107-114`.
 */
fun materialize(deviceId: DeviceId, ops: List<Operation>): Map<ItemId, SyncItem> {
    val sorted = ops.sortedBy { it.timestamp }
    return SyncState.empty(deviceId).applyAll(sorted).items
}

/**
 * Convert a materialized sync state to a flat list of [Todo] values,
 * skipping deleted items. Matches `materializeToTodos` at
 * `src/Todo/Sync/CRDT.hs:117-122`.
 *
 * The returned items use the LWW values for priority and
 * description, not the snapshot stored in [SyncItem.item] — that's
 * what lets remote edits to those fields show through after a sync.
 */
fun materializeToTodos(items: Map<ItemId, SyncItem>): List<Todo> =
    items.values
        .filter { !it.deleted }
        .map { it.toTodo() }

/** Convert a single [SyncItem] to a [Todo], folding the LWW registers back into the item. */
fun SyncItem.toTodo(): Todo {
    val merged = item.copy(
        priority = priority.value,
        description = description.value,
    )
    return if (completed) Todo.Completed(merged) else Todo.Incomplete(merged)
}
